### PR TITLE
DD-32: Payment reminder activity

### PR DESCRIPTION
**Pull request by awg-github**
1. In "Direct Debit Configurations" under "Payment Config" section, a number field "Days in advance for Collection Reminder" is provided. This field is provided with help text "Set how many days in advance of the Received Date of a Direct Debit contribution should the Direct Debit Payment Collection Reminder be sent to the user."

2. A scheduled job "Send Direct Debit Payment Collection Reminders" is created. It's default frequency  is "Daily".

3. Every time the "Send Direct Debit Payment Collection Reminders" job is executed, it will find any contribution that meets all criteria below:
- Payment Method is Direct Debit
- Contribution Status is Pending or Cancelled
- Receipt Date is empty
- "Received Date" minus "Days in advance for Collection Reminder" is equal to or smaller than today

4. For each contribution in point 3, the scheduled job will generate a "Direct Debit Payment Collection Reminder" activity with:
- Subject - "Direct Debit Payment Collection Reminder"
- Source Record ID - ID of the Contribution
- Added by - Admin
- With - The user

dd-32 payment reminder activity

### CODE REVIEW

File: CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        ) AS email
+      FROM civicrm_contribution AS contribution
+      LEFT JOIN civicrm_contact AS contact
+        ON contribution.contact_id = contact.id
+      WHERE 
+        contribution.receipt_date IS NULL
+        AND contribution.payment_instrument_id = %2
+        AND (contribution.contribution_status_id = %3 OR contribution.contribution_status_id = %4)
+        AND contribution.receive_date <= (now() - INTERVAL %1 DAY)
+    ";
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php]: Two things to have into account:

- First, we should get the primary e-mail for the contact.
- Second, we should take into account if the contact has opted out of e-mails, bearing in mind there are two flags to opt-out, so I don't know if one of both should be evaluated (opt_out vs no_email). You may want to run this one by Guanhuan to verify if and which of the flags should be read.

File: CRM/ManualDirectDebit/Upgrader.php [hunk 2]
```diff
@@ -226,11 +250,23 @@ public function onEnable() {
...
-    $this->deleteMessageTemplates();
+  }
+
+  /**
+   * Deletes scheduled job created by the extension
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function deleteScheduledJob() {
+    civicrm_api3('Job', 'get', [
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Upgrader.php]: We also need to implement onEnable and onDisable events to handle what happens with the scheduled job. Basically, we need to set the `is_active` flag to true on enable, false on disable.

File: CRM/ManualDirectDebit/ScheduleJob/Reminder.php [hunk 3]
```diff
@@ -40,7 +40,7 @@ public function run() {
         $this->sendEmail($targetContributionData['contributionId']);
       }
       else {
-        $this->setLog(ts("Email not sent. Related contact haven't email."));
+        $this->setLog(ts("Email not sent. Related contact haven't email or turn on 'do not send email'"));
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/Reminder.php]: Please change the message to:

> Email not sent. Related contact doesn't have an e-mail or has the 'do not send e-mail' flag set.

File: CRM/ManualDirectDebit/ScheduleJob/ReminderHandler.php [hunk 4]
```diff
@@ -23,6 +23,7 @@ public function init() {
     if (!$this->hasRequirements()) {
       $log['is_error'] = 1;
       $log['values'] = $this->requirementsErrorMessage;
+      $log['error_message'] = $this->requirementsErrorMessage;
       return $log;
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/ReminderHandler.php]: Blank line before return.

File: CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php [hunk 5]
```diff
@@ -24,6 +24,7 @@ public static function retrieve() {
           SELECT email.email
           FROM civicrm_email AS email
           WHERE email.contact_id = contact.id
+            AND (contact.do_not_email IS NULL OR contact.do_not_email = 0)
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php]: We should also use the is_primary flag, to be sure to send to primary e-mail.

```sql
SELECT email.email
FROM civicrm_email AS email
WHERE email.contact_id = contact.id
AND (contact.do_not_email IS NULL OR contact.do_not_email = 0)
AND is_primary = 1
```

File: CRM/ManualDirectDebit/Upgrader.php [hunk 6]
```diff
@@ -180,6 +187,25 @@ private function deleteMessageTemplates() {
...
+    $domainID = CRM_Core_Config::domainID();
+
+    $params = [
+      'name' => 'Send Direct Debit Payment Collection Reminders',
+      'description' => 'Send Direct Debit Payment Collection Reminders',
+      'api_entity' => 'ManualDirectDebit',
+      'api_action' => 'run',
+      'run_frequency' => 'Daily',
+      'domain_id' => $domainID,
+      'is_active' => '0',
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Upgrader.php]: Are you sure the scheduled job should start inactive?

On that note, we also need to implement onDisable and onEnable for the scheduled job too. On each case, the 'is_active' flag should be set to true or false, correspondingly.

Comment (awg-github) [reply CRM/ManualDirectDebit/Upgrader.php]: We have already implemented onDisable() and onEnable() for the scheduled job. Look at the method 'alterCustomValues()'  in Upgrader.

Comment (MiyaNoctem) [reply CRM/ManualDirectDebit/Upgrader.php]: Oh, ok, got it now. Still, I think it is a problem it wasn't obvious that method also takes care of enabling/disabling the scheduled job. So this is my suggestion of how we can improve this:

1. Separate CRM_ManualDirectDebit_Upgrader::$customValues into different class attributes (eg. $customOptionValues, $scheduledJobs, $processorTyes, $optionGroups).
2. Be sure to add a doc-block to each new attribute. 
3. Rename `alterCustomValues()` to `alterEntitiesValues()`.
4. Refactor `alterEntitiesValues()` to something like:
```php
private function alterCustomEntityValues($action) {
  $entities = array_merge($this->optionGroups, $this->scheduledJobs, $this->processorTypes, $this->optionValues);

  foreach ($entities as $customEntityValue) {
    $this->alterEntity(
      $customEntityValue['entityType'],
      $customEntityValue['searchValue'],
      isset($customEntityValue['searchField']) ? $customEntityValue['searchField'] : 'name',
      isset($customEntityValue['optionGroup']) ? $customEntityValue['optionGroup'] : FALSE,
      $action
    );
  }
}
```

File: CRM/ManualDirectDebit/ScheduleJob/Reminder.php [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    foreach ($targetContributionDataList as $targetContributionData) {
+      $this->setLog(ts("Contribution with id = %1", [1 => $targetContributionData['contributionId']]));
+
+      if (!empty($targetContributionData['email'])) {
+        $this->sendEmail($targetContributionData['contributionId']);
+      }
+      else {
+        $this->setLog(ts("Email not sent. Related contact haven't email or turn on 'do not send email'"));
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/Reminder.php]: Please change the message to:

> Email not sent. Related contact doesn't have an e-mail or has the 'do not send e-mail' flag set.

File: CRM/ManualDirectDebit/ScheduleJob/ReminderHandler.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+   * @return array
+   */
+  public function init() {
+    $log = [];
+
+    if (!$this->hasRequirements()) {
+      $log['is_error'] = 1;
+      $log['values'] = $this->requirementsErrorMessage;
+      $log['error_message'] = $this->requirementsErrorMessage;
+      return $log;
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/ReminderHandler.php]: Blank line before return.

File: CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+    $query = "
+      SELECT 
+        contribution.id AS contribution_id,
+        contact.id AS contact_id,
+        (
+          SELECT email.email
+          FROM civicrm_email AS email
+          WHERE email.contact_id = contact.id
+            AND (contact.do_not_email IS NULL OR contact.do_not_email = 0)
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/ScheduleJob/TargetContribution.php]: We should also use the is_primary flag, to be sure to send to primary e-mail.

```sql
SELECT email.email
FROM civicrm_email AS email
WHERE email.contact_id = contact.id
AND (contact.do_not_email IS NULL OR contact.do_not_email = 0)
AND is_primary = 1
```

**Pull request merged**